frontend/lib/providers/artefact_builds.dart

```diff
@@ -19,22 +19,47 @@ class ArtefactBuilds extends _$ArtefactBuilds {
     return artefactBuilds;
   }
 
+  void _updateStateNewTestExecution(
+    int testExecutionId,
+    Map<String, Object?> responseData,
+  ) {
+    final List<ArtefactBuild> newStateData = [
+      for (final artefactBuild in state.value ?? [])
+        ArtefactBuild(
+          id: artefactBuild.id,
+          architecture: artefactBuild.architecture,
+          revision: artefactBuild.revision,
+          testExecutions: List.from(
+            artefactBuild.testExecutions.map(
+              (e) {
+                if (e.id == testExecutionId) {
+                  return TestExecution.fromJson(responseData);
+                }
+                return e;
+              },
+            ),
+          ),
+        ),
```

Comment:
I don't think that this code is completely correct because at doesn't make copies of other test executions and doesn't copy environments. And state [should be immutable](https://riverpod.dev/docs/concepts/why_immutability#using-immutable-state). I recommend using `copyWith` for artefact builds and test executions and possibly adding a helper function.

Reply:
I see, this code is indeed not really good idea, but I think using copyWith wont be much simpler, as we again need to rebuild at least one artefact build. It seems like a better option is to make that additional network request after all..

Reply:
adding a network request feels like choosing a worst solution for the sake of better code. We don't need to rebuild one artefact build. Once we copyWith everything, we just modify one artefact build

Reply:
Oh ignore my point about modifying one artefact build as it's a freezed object so it can't be modified

Reply:
I agree with your points, I think my implementation should be updating the state correctly now.